heron/metricscachemgr/src/java/com/twitter/heron/metricscachemgr/MetricsCacheManagerHttpServer.java

```diff
@@ -0,0 +1,157 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.twitter.heron.metricscachemgr;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.util.logging.Logger;
+
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.sun.net.httpserver.HttpExchange;
+import com.sun.net.httpserver.HttpHandler;
+import com.sun.net.httpserver.HttpServer;
+
+import com.twitter.heron.metricscachemgr.metricscache.MetricsCache;
+import com.twitter.heron.proto.tmaster.TopologyMaster;
+import com.twitter.heron.spi.utils.NetworkUtils;
+
+
+/**
+ * MetricsCacheMgr http server:
+ * compatible with tmaster and tracker http interface for metrics
+ * http path:
+ * "/stats" metric query
+ * "/exceptions" exception query
+ * "/exceptionsummary" exception query, with aggregation
+ * <p>
+ * Differece from MetricsCacheManagerServer:
+ * 1. MetricsCacheManagerServer accepts metric publishing message from sinks;
+ * MetricsCacheManagerHttpServer responds to queries.
+ * 2. MetricsCacheManagerServer is a HeronServer;
+ * MetricsCacheManagerHttpServer is a http server
+ */
+public class MetricsCacheManagerHttpServer {
+  // http path, compatible with tmaster stat interface
+  public static final String PATH_STATS = "/stats";
+  public static final String PATH_EXCEPTIONS = "/exceptions";
+  public static final String PATH_EXCEPTIONSUMMARY = "/exceptionsummary";
+  // logger
+  private static final Logger LOG = Logger.getLogger(MetricsCacheManagerHttpServer.class.getName());
+  // http server
+  private HttpServer server = null;
+  // reference to MetricsCache object
+  private MetricsCache metricsCache = null;
+
+  public MetricsCacheManagerHttpServer(MetricsCache cache, int port) throws IOException {
+    metricsCache = cache;
+
+    server = HttpServer.create(new InetSocketAddress(port), 0);
+
+    server.createContext(PATH_STATS, new HandleStatsRequest());
+    server.createContext(PATH_EXCEPTIONS, new HandleExceptionRequest());
+    server.createContext(PATH_EXCEPTIONSUMMARY, new HandleExceptionSummaryRequest());
+  }
+
+  public void start() {
+    server.start();
+  }
+
+  public void stop() {
+    server.stop(0);
+  }
+
+  // compatible with tmaster stat interface: http+protobuf
+  public class HandleStatsRequest implements HttpHandler {
+
+    @Override
+    public void handle(HttpExchange httpExchange) throws IOException {
+      // get the entire stuff
+      byte[] payload = NetworkUtils.readHttpRequestBody(httpExchange);
+      TopologyMaster.MetricRequest req;
+      try {
+        req = TopologyMaster.MetricRequest.parseFrom(payload);
+      } catch (InvalidProtocolBufferException e) {
+        LOG.severe("Unable to decipher data specified in StatsRequest " + e.toString());
```

Comment:
Typically you don't want to log `e.toString()` but instead pass `e` to the logger like this:
```
LOG.log(Level.SEVERE, "Unable to decipher data specified in StatsRequest: " + httpExchange, e);
```
Replace httpExchange with anything else that provides helpful info about the uri and/or the payload. Same comment for other handlers in this class.